src/main/cpp/io/ArmIO.cpp

```diff
@@ -32,23 +32,83 @@ bool ArmIO::ProcessIO() {
         sw_interface_->reset_encoders = false;
     }
 
-    
-
-    // Get the hardware sensor values.
-    // limit switches
-    // sw_interface_->deployed_limit_switch_val =
-    //     hw_interface_->deploy_limit_switch->Get(); //???
-    // sw_interface_->retracted_limit_switch_val =
-    //     hw_interface_->retracted_limit_switch->Get(); //???
-
     // intake position encoder
     OKC_CHECK(hw_interface_->arm_extend_motor != nullptr);
     OKC_CHECK(hw_interface_->arm_up_motor != nullptr);
-    hw_interface_->arm_lift_motor->Set(sw_interface_->arm_lift_power);
-    hw_interface_->arm_up_motor->Set(sw_interface_->arm_up_power);
-    hw_interface_->arm_extend_motor->Set(sw_interface_->arm_extend_power);
+   
+    sw_interface_->arm_encoder = hw_interface_->arm_lift_encoder->GetPosition();
+    sw_interface_->arm_extend_encoder = hw_interface_->arm_extend_encoder->GetPosition();
+    sw_interface_->arm_duty_cycle_encoder = hw_interface_->arm_duty_cycle_encoder->GetAbsolutePosition() * 360  -  330; // offset of 330
+
+    if (sw_interface_->arm_duty_cycle_encoder < 0) {
+        sw_interface_->arm_duty_cycle_encoder += 360;
+        sw_interface_->arm_duty_cycle_encoder *= -1;
+    }
+    
+    TeamOKC::Clamp(-0.2, 0.2, &sw_interface_->arm_lift_power);
+    TeamOKC::Clamp(-0.2, 0.2, &sw_interface_->arm_extend_power);
+
+    // if the absolute encoder is >110 degrees
+    if (sw_interface_->arm_duty_cycle_encoder >= 110) {
+        // and we're trying to go farther positive
+        if (sw_interface_->arm_lift_power > 0) {
+            // stop it
+            hw_interface_->arm_lift_motor->Set(0);
+            hw_interface_->arm_up_motor->Set(0);
+        } else {
+            // otherwise, it's cool
+            hw_interface_->arm_lift_motor->Set(sw_interface_->arm_lift_power);
+            hw_interface_->arm_up_motor->Set(-sw_interface_->arm_lift_power);
+        }
+    } else if (sw_interface_->arm_duty_cycle_encoder <= -110) {
+        // and we're trying to go farther negative
+        if (sw_interface_->arm_lift_power < 0) {
+            // stop it
+            hw_interface_->arm_lift_motor->Set(0);
+            hw_interface_->arm_up_motor->Set(0);
+        } else {
+            // otherwise, it's cool
+            hw_interface_->arm_lift_motor->Set(sw_interface_->arm_lift_power);
+            hw_interface_->arm_up_motor->Set(-sw_interface_->arm_lift_power);
+        }
+    } else { // if neither limit is reached
+        hw_interface_->arm_lift_motor->Set(sw_interface_->arm_lift_power);
+        hw_interface_->arm_up_motor->Set(-sw_interface_->arm_lift_power);
+    }
```

Comment:
On one hand this is a great use of the IO class because it guarantees hardware protection. On the other hand, the software subsystem is probably winding up the integrator of the PID if these limits are being hit. 

No change needed yet, but if control issues start appearing with a non-zero Ki, then we will need to send a boolean flag up to the software to tell it to stop adding to the integrator

Reply:
I don't think we'll end up using an I-term, but yes. I think we should also do some PID setpoint limiting to the actual hardware (and soft) limits, because I was noticing that holding down the arm buttons can get your setpoint really high past the limits really quickly, which isn't good. I might write something in Mattermost up.